File: IOI/2007/flood/solutions/aliens.pas
program aliens;

var
   n, x0, y0 : longint;
   left, right, up : longint;
   sideLength : longint;
   squaresLeft, squaresRight, squaresUp : longint;
   edgeLeft, edgeUp : longint;

function Rose( X, Y : longint ) : boolean;
var
   answer : string;
begin
   writeln( 'examine ', X, ' ', Y );
   flush( output );
   readln( answer );
   if answer = 'true' then Rose := true else Rose := false;
end;

function query( x, y : longint ) : boolean;
begin
   if (x < 1) or (x > n) or (y < 1) or (y > n) then query := false
   else query := Rose(x, y);
end;

function findFirstWhite( dx, dy : longint ) : longint;
var
   p, lo, hi, mid : longint;
begin
   p := 0;
   while query(x0+(dx shl p)-dx, y0+(dy shl p)-dy) do p := p + 1;

   lo := (1 shl (p-1))-1;
   hi := (1 shl p)-1;
   while lo+1 < hi do begin
      mid := (lo+hi) div 2;
      if query(x0+mid*dx, y0+mid*dy) then lo := mid else hi := mid;
   end;

   findFirstWhite := hi;
end;


begin;
   readln( n, x0, y0 );

   left := findFirstWhite(0, -1);
   right := findFirstWhite(0, 1);
   up := findFirstWhite(-1, 0);

   sideLength := right+left-1;
   squaresLeft := findFirstWhite(0, -2*sideLength);
   squaresRight := findFirstWhite(0, 2*sideLength);
   squaresUp := findFirstWhite(-2*sideLength, 0);

   edgeLeft := y0-left+1-(squaresLeft-1)*2*sideLength;
   edgeUp := x0-up+1-(squaresUp-1)*2*sideLength;

   if squaresLeft+squaresRight-1 <> 3 then begin;
      edgeLeft := edgeLeft - sideLength;
      edgeUp := edgeUp - sideLength;
   end;

   writeln( 'solution ', edgeUp+5*sideLength div 2, ' ', edgeLeft+5*sideLength div 2 );
   flush( output );
end.

File: IOI/2007/flood/solutions/aliens.cpp
#include <map>
#include <algorithm>

#include <cstdio>
#include <cstring>
#define SOLUTIONCOMMAND		"solution"
#define EXAMINECOMMAND		"examine"
void Initialize ( int *N, int *x0, int *y0 )
{
	scanf ("%d%d%d", N, x0, y0); 
}

int Rose ( int X, int Y )
{
	printf( "%s %d %d\n", EXAMINECOMMAND, X, Y );
	fflush( stdout );

	char answer[10];
	scanf ("%s", answer);

	return strcmp(answer, "true") == 0;
}

void Solution ( int X, int Y )
{
	printf( "%s %d %d\n", SOLUTIONCOMMAND, X, Y );
	fflush( stdout );
}


using namespace std;

int n, x0, y0;


int query(int x, int y) {
  if (x < 1 || x > n || y < 1 ||  y > n)
    return 0;
  return Rose(x, y);
}


int find_first_white(int dx, int dy) {
  int i = 0;
  while (query(x0+(dx<<i)-dx, y0+(dy<<i)-dy))
    i++;

  int lo = (1<<(i-1))-1;
  int hi = (1<<i)-1;
  while (lo+1 < hi) {
    int mid = (lo+hi)/2;
    if (query(x0+mid*dx, y0+mid*dy))
      lo = mid;
    else
      hi = mid;
  }
  return hi;
}


int main() {
  Initialize(&n, &x0, &y0);

  int left = find_first_white(0, -1);
  int right = find_first_white(0, 1);
  int up = find_first_white(-1, 0);

  int side_length = right+left-1;
  int squares_left = find_first_white(0, -2*side_length);
  int squares_right = find_first_white(0, 2*side_length);
  int squares_up = find_first_white(-2*side_length, 0);

  int edge_left = y0-left+1-(squares_left-1)*2*side_length;
  int edge_up = x0-up+1-(squares_up-1)*2*side_length;

  if (squares_left+squares_right-1 != 3) {
    edge_left -= side_length;
    edge_up -= side_length;
  }

  Solution(edge_up+5*side_length/2, edge_left+5*side_length/2);

  return 0;
}
